portfolio/src/main/java/com/google/sps/data/WordCountUpdater.java

```diff
@@ -0,0 +1,72 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.data;
+
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.Transaction;
+import com.google.appengine.api.datastore.TransactionOptions;
+
+/** Class that handles the updates of word counts upon addition or deletion of comment. */
+public class WordCountUpdater {
+
+  private static final int TRANSACTION_RETRIES = 4;
+  public static enum UpdateOp {
+    ADD_WORDS,
+    REMOVE_WORDS
+  }
+  private DatastoreService datastore;
+  
+  public WordCountUpdater() {
+    datastore = DatastoreServiceFactory.getDatastoreService();
+  }
+  
+  public void updateWordCount(String comment, UpdateOp op) {
+    String[] words = comment.replaceAll("[^a-zA-Z ]", "").toLowerCase().split("\\s+");
+    
+    for (String word : words) {
+      Entity wordEntity;
+      Key wordKey = KeyFactory.createKey("word", word + "_word");
+      Transaction txn = datastore.beginTransaction(TransactionOptions.Builder.withXG(true));
+      try {
+        wordEntity = datastore.get(txn, wordKey);
+      } catch (com.google.appengine.api.datastore.EntityNotFoundException e) {
+        // In case word is not yet in datastore...
+        wordEntity = new Entity(wordKey);
+        wordEntity.setProperty("count", 0L);
+      }
+      long newCount = (Long) wordEntity.getProperty("count") + (op == UpdateOp.ADD_WORDS ? 1 : -1);
```

Comment:
Nit: You should really match enums (and the string, before) exhaustively, rather than assuming that there will always only be two variants.

It doesn't matter in this toy context, but in larger applications, this structure of the code would allow someone to silently add a new variant with the code would continuing to compile. It would no longer be correct though.

I know it's a bit more verbose, but consider the explicit switch here.